blocks/loops.js

```diff
@@ -20,165 +20,252 @@
-      var TOOLTIPS = {
-        'WHILE': Blockly.Msg.CONTROLS_WHILEUNTIL_TOOLTIP_WHILE,
-        'UNTIL': Blockly.Msg.CONTROLS_WHILEUNTIL_TOOLTIP_UNTIL
-      };
-      return TOOLTIPS[op];
-    });
-  }
+Blockly.Extensions.register('controls_whileUntil_tooltip',
+  Blockly.Extensions.buildTooltipForDropdown(
+    'MODE', Blockly.Constants.Loops.WHILE_UNTIL_TOOLTIPS));
+
+/**
+ * Tooltips for the 'controls_flow_statements' block, keyed by FLOW value.
+ * Messages references are replaced just before displayed.
+ * @package
+ * @readonly
+ */
+Blockly.Constants.Loops.BREAK_CONTINUE_TOOLTIPS = {
+  'BREAK': '%{BKY_CONTROLS_FLOW_STATEMENTS_TOOLTIP_BREAK}',
+  'CONTINUE': '%{BKY_CONTROLS_FLOW_STATEMENTS_TOOLTIP_CONTINUE}'
 };
 
-Blockly.Blocks['controls_for'] = {
-  /**
-   * Block for 'for' loop.
-   * @this Blockly.Block
-   */
-  init: function() {
-    this.jsonInit({
-      "message0": Blockly.Msg.CONTROLS_FOR_TITLE,
-      "args0": [
-        {
-          "type": "field_variable",
-          "name": "VAR",
-          "variable": null
-        },
-        {
-          "type": "input_value",
-          "name": "FROM",
-          "check": "Number",
-          "align": "RIGHT"
-        },
-        {
-          "type": "input_value",
-          "name": "TO",
-          "check": "Number",
-          "align": "RIGHT"
-        },
-        {
-          "type": "input_value",
-          "name": "BY",
-          "check": "Number",
-          "align": "RIGHT"
-        }
-      ],
-      "inputsInline": true,
-      "previousStatement": null,
-      "nextStatement": null,
-      "colour": Blockly.Blocks.loops.HUE,
-      "helpUrl": Blockly.Msg.CONTROLS_FOR_HELPURL
-    });
-    this.appendStatementInput('DO')
-        .appendField(Blockly.Msg.CONTROLS_FOR_INPUT_DO);
-    // Assign 'this' to a variable for use in the tooltip closure below.
-    var thisBlock = this;
-    this.setTooltip(function() {
-      return Blockly.Msg.CONTROLS_FOR_TOOLTIP.replace('%1',
-          thisBlock.getFieldValue('VAR'));
-    });
-  },
+Blockly.Extensions.register('controls_flow_tooltip',
+  Blockly.Extensions.buildTooltipForDropdown(
+    'FLOW', Blockly.Constants.Loops.BREAK_CONTINUE_TOOLTIPS));
+
+/**
+ * Mixin to add a context menu item to create a 'variables_get' block.
+ * Used by blocks 'controls_for' and 'controls_forEach'.
+ * @mixin
+ * @augments Blockly.Block
+ * @package
+ * @readonly
+ */
+Blockly.Constants.Loops.CUSTOM_CONTEXT_MENU_CREATE_VARIABLES_GET_MIXIN = {
   /**
-   * Add menu option to create getter block for loop variable.
+   * Add context menu option to create getter block for the loop's variable.
+   * (customContextMenu support limited to web BlockSvg.)
    * @param {!Array} options List of menu options to add to.
    * @this Blockly.Block
    */
-  customContextMenu: function(options) {
-    if (!this.isCollapsed()) {
+  customContextMenu: function() {
+    var varName = this.getFieldValue('VAR');
+    if (!this.isCollapsed() && varName != null) {
```

Comment:
Scratch team has been discussing having an empty option for no selected variable. +1 to checking just in case.